packages/ckeditor5-link/tests/autolink.js

```diff
@@ -0,0 +1,225 @@
+			);
+		} );
+
+		it( 'adds linkHref attribute to a text link after a soft break', () => {
+			setData( model, '<paragraph>https://www.cksource.com[]</paragraph>' );
+
+			editor.execute( 'shiftEnter' );
+
+			// TODO: should test with selection but master has a bug. See: https://github.com/ckeditor/ckeditor5/issues/7459.
+			expect( getData( model, { withoutSelection: true } ) ).to.equal(
+				'<paragraph>' +
+					'<$text linkHref="https://www.cksource.com">https://www.cksource.com</$text>' +
+					'<softBreak></softBreak>' +
+				'</paragraph>'
+			);
+		} );
+
+		it( 'does not add linkHref attribute to a text link after double soft break', () => {
+			setData( model, '<paragraph>https://www.cksource.com<softBreak></softBreak>[]</paragraph>' );
+
+			editor.execute( 'shiftEnter' );
+
+			expect( getData( model ) ).to.equal(
+				'<paragraph>https://www.cksource.com<softBreak></softBreak><softBreak></softBreak>[]</paragraph>'
+			);
+		} );
+
+		it( 'adds linkHref attribute to a text link on enter', () => {
+			setData( model, '<paragraph>https://www.cksource.com[]</paragraph>' );
+
+			editor.execute( 'enter' );
+
+			expect( getData( model ) ).to.equal(
+				'<paragraph>' +
+					'<$text linkHref="https://www.cksource.com">https://www.cksource.com</$text>' +
+				'</paragraph>' +
+				'<paragraph>[]</paragraph>'
+			);
+		} );
+
+		it( 'adds "mailto://" to link of detected email addresses', () => {
+			simulateTyping( 'newsletter@cksource.com ' );
+
+			expect( getData( model ) ).to.equal(
+				'<paragraph><$text linkHref="mailto://newsletter@cksource.com">newsletter@cksource.com</$text> []</paragraph>'
+			);
+		} );
+
+		const supportedURLs = [
+			'http://cksource.com',
+			'https://cksource.com',
+			'http://www.cksource.com',
+			'hTtP://WwW.cKsOuRcE.cOm',
+			'http://foo.bar.cksource.com',
+			'http://www.cksource.com/some/path/index.html#abc',
+			'http://www.cksource.com/some/path/index.html?foo=bar',
+			'http://www.cksource.com/some/path/index.html?foo=bar#abc',
+			'http://localhost',
+			'ftp://cksource.com',
+			'mailto://cksource@cksource.com',
+			'www.cksource.com',
+			'cksource.com'
+		];
+
+		const unsupportedURLs = [
+			'http://www.cksource.com/some/path/index.html#abc?foo=bar', // Wrong #? sequence.
+			'http:/cksource.com'
+		];
+
+		for ( const supportedURL of supportedURLs ) {
+			it( `should detect "${ supportedURL }" as a valid URL`, () => {
+				simulateTyping( supportedURL + ' ' );
+
+				expect( getData( model ) ).to.equal(
+					`<paragraph><$text linkHref="${ supportedURL }">${ supportedURL }</$text> []</paragraph>` );
+			} );
+		}
+
+		for ( const unsupportedURL of unsupportedURLs ) {
+			it( `should not detect "${ unsupportedURL }" as a valid URL`, () => {
+				simulateTyping( unsupportedURL + ' ' );
+
+				expect( getData( model ) ).to.equal(
+					`<paragraph>${ unsupportedURL } []</paragraph>` );
+			} );
+		}
+	} );
```

Comment:
I miss a test where you check what will happen after typing a text. Not a link, not an email. Just a text.

Reply:
Done.